ports/geckolib/gecko_bindings/sugar/ownership.rs

```diff
@@ -0,0 +1,340 @@
+/// use HasArcFFI or similar traits to define it
+pub unsafe trait HasFFI : Sized {
+    type FFIType: Sized;
+}
+
+/// Indicates that a given Servo type has the same layout
+/// as the corresponding HasFFI::FFIType type
+pub unsafe trait HasSimpleFFI : HasFFI {
+    #[inline]
+    /// Given a Servo-side reference, converts it to an
+    /// FFI-safe reference which can be passed to Gecko
+    ///
+    /// &ServoType -> &GeckoType
+    fn as_ffi(&self) -> &Self::FFIType {
+        unsafe { transmute(self) }
+    }
+    #[inline]
+    /// Given a Servo-side mutable reference, converts it to an
+    /// FFI-safe mutable reference which can be passed to Gecko
+    ///
+    /// &mut ServoType -> &mut GeckoType
+    fn as_ffi_mut(&mut self) -> &mut Self::FFIType {
+        unsafe { transmute(self) }
+    }
+    #[inline]
+    /// Given an FFI-safe reference obtained from Gecko
+    /// converts it to a Servo-side reference
+    ///
+    /// &GeckoType -> &ServoType
+    fn from_ffi(ffi: &Self::FFIType) -> &Self {
+        unsafe { transmute(ffi) }
+    }
+    #[inline]
+    /// Given an FFI-safe mutable reference obtained from Gecko
+    /// converts it to a Servo-side mutable reference
+    ///
+    /// &mut GeckoType -> &mut ServoType
+    fn from_ffi_mut(ffi: &mut Self::FFIType) -> &mut Self {
+        unsafe { transmute(ffi) }
+    }
+}
+
+/// Indicates that the given Servo type is passed over FFI
+/// as a Box
+pub unsafe trait HasBoxFFI : HasSimpleFFI {
+    #[inline]
+    fn into_ffi(self: Box<Self>) -> Owned<Self::FFIType> {
+        unsafe { transmute(self) }
+    }
+}
+
+/// Helper trait for conversions between FFI Strong/Borrowed types and Arcs
+///
+/// Should be implemented by types which are passed over FFI as Arcs
+/// via Strong and Borrowed
+///
+/// In this case, the FFIType is the rough equivalent of ArcInner<Self>
+pub unsafe trait HasArcFFI : HasFFI {
+    // these methods can't be on Borrowed because it leads to an unspecified
+    // impl parameter
+    /// Artificially increments the refcount of a borrowed Arc over FFI.
+    unsafe fn addref(ptr: Borrowed<Self::FFIType>) {
+        forget(ptr.as_arc_opt::<Self>().clone())
+    }
+
+    /// Given a (possibly null) borrowed FFI reference, decrements the refcount.
+    /// Unsafe since it doesn't consume the backing Arc. Run it only when you
+    /// know that a strong reference to the backing Arc is disappearing
+    /// (usually on the C++ side) without running the Arc destructor.
+    unsafe fn release(ptr: Borrowed<Self::FFIType>) {
+        if let Some(arc) = ptr.as_arc_opt::<Self>() {
+            let _: Arc<_> = ptr::read(arc as *const Arc<_>);
+        }
+    }
+}
+
+#[repr(C)]
+/// Gecko-FFI-safe borrowed type
+/// This can be null.
+pub struct Borrowed<'a, T: 'a> {
+    ptr: *const T,
+    _marker: PhantomData<&'a T>,
+}
+
+#[repr(C)]
+/// Gecko-FFI-safe mutably borrowed type
+/// This can be null.
+pub struct BorrowedMut<'a, T: 'a> {
+    ptr: *mut T,
```

Comment:
Seems to me that this could be `Option<*mut T>`? Then most types could use `as_ref()`, or similar.